plasmapy/simulation/particle_tracker.py

```diff
@@ -0,0 +1,835 @@
+        save_routine: Optional[AbstractSaveRoutine] = None,
+        dt=None,
+        field_weighting="volume averaged",
+    ):
+        r"""
+        Runs a particle-tracing simulation.
+        Timesteps are adaptively calculated based on the
+        local grid resolution of the particles and the electric and magnetic
+        fields they are experiencing.
+
+        Parameters
+        ----------
+        dt : `~astropy.units.Quantity`, optional
+            An explicitly set timestep in units convertible to seconds.
+            Setting this optional keyword overrules the adaptive time step
+            capability and forces the use of this timestep throughout. If a tuple
+            of timesteps is provided, the adaptive timestep will be clamped
+            between the first and second values.
+
+        field_weighting : str
+            String that selects the field weighting algorithm used to determine
+            what fields are felt by the particles. Options are:
+
+            * 'nearest neighbor': Particles are assigned the fields on
+                the grid vertex closest to them.
+            * 'volume averaged' : The fields experienced by a particle are a
+                volume-average of the eight grid points surrounding them.
+
+            The default is 'volume averaged'.
+
+        stop_condition : callable
+            A function responsible for determining when the simulation has reached
+            a suitable termination condition. The function is passed an instance
+            of `~plasmapy.simulation.particle_tracker.ParticleTracker`. The function
+            must return a tuple of a boolean, corresponding to the completion status of
+            the simulation, and a number, representing the progress of the simulation.
+
+        Returns
+        -------
+        None
+
+        """
+
+        # Validate inputs to the run function
+        # Sets is_uniform_time property
+        self._validate_run_inputs(stop_condition, save_routine, dt, field_weighting)
+        self._enforce_particle_creation()
+
+        # By default, set dt as an infinite range (auto dt with no restrictions)
+        self.dt = np.array([0.0, np.inf]) * u.s if dt is None else dt
+        self.dt = (self.dt).to(u.s).value
+
+        # Keep track of how many push steps have occurred for trajectory tracing
+        self.iteration_number = 0
+
+        self.time = 0 if self.is_uniform_time else np.zeros((self.nparticles, 1))
+        # Create flags for tracking when particles during the simulation
+        # on_grid -> zero if the particle is off grid, 1
+        # shape [nparticles, ngrids]
+        self.on_grid = np.zeros([self.nparticles, self.num_grids])
+
+        # Entered grid -> non-zero if particle EVER entered a grid
+        self.entered_grid = np.zeros([self.nparticles])
+
+        # Update the `particle_tracker` attribute so that the stop condition & save routine can be used
+        stop_condition.particle_tracker = self
```

Comment:
Here you are setting the `particle_tracker` attribute manually, but at least the save routine has a setter method. I think maybe it's best to add a setter method to the stop condition as well, then use those here? That will make it easier to document that this has to be done.